documents/readthedoc/docs/source/Solutions/ehsm-kms/ehsm.md

```diff
@@ -0,0 +1,88 @@
+---
+
+## Introduction to Keys used in eHSM-KMS
+
+eHSM-KMS will provide different kinds of keys, but none of them will be exposed outside of the SGX enclave.
+The below diagram depicts the overview of the keys used in the eHSM-KMS,
+![keys-in-ehsm-kms](./img/keys-in-ehsm-kms.PNG)
+
+ - DK (DomainKey) is an AES key that is used to protect/wrap all the CMKs in a domain (region).
+    - It must be securely distributed/provisioned cross all SGX-capable machines (or nodes) in a domain.
+ - CMK (Customer Master Key) could be symmetric or asymmetric keys.
+    - Symmetric CMK mainly used to wrap the DataKey, also can be used to encrypted an arbitrary set of bytes data(<6KB).
+    - Asymmetric CMK mainly used to sign/verify or asymmetric encrypt/decrypt data (not for the DataKey.)
+ - DataKey is a symmetric key with random size of bytes.
+    - The plaintext of the DataKey is used to encrypt data locally, which should be cleared from memory once been used as soon as possible.
+    - When a user wants to obtain the plaintext of DataKey again, it needs to call the Decrypt with the CMK to get the plaintext of DataKey.
+
+
+## Domain and Worker Nodes
+The below diagram shows the definitions of the domain and worker nodes in the eHSM-KMS,
+    ![domain-and-nodes](./img/domain-and-nodes.PNG)
+
+  - All the SGX Nodes (SGX-capable platforms) in a Domain can equivalently & seamlessly provide KMS services for users. 
+
+  - eHSM-core Enclave in each node should be provisioned with the capability of retrieving the cleartext of DomainKey. (DomainKey sharing across all the members in one domain)
+  - Each node may be offline/down. New node may be joined to or removed from a specific domain.
+
+---
+
+## DomainKey Provisioning
+eHSM-KMS provides a protocol based the SGX attestation to securely provision the DomainKey to each eHSM-core Enclave which is illustrated in the following diagram.
+![domainkey-provisioning](./img/domainkey-provisioning.PNG)
+
+ - The centralized deployment node could be an HSM-equipped platform or an SGX-capable platform.
+ - eHSM-dkeyserver runs as a TCP/IP socket server, and it’s responsible for generating DK and protecting it in the HSM or enclave and provisioning it to each SGX members in this domain.
+ - eHSM-dkeycache runs as a proxy agent in each SGX Node (only need one instance for each node). And it’s responsible for retrieving the DK from the ehsm-dkeyserver via remote secure channel and then distributing to each ehsm-core instance via local secure channel.
+ - Whenever a new node is joined or refreshed into the domain, the ehsm-dkeycache service will try to connect the ehsm-dkeyserver to retrieve the DK and distribute it to each ehsm-core instances via the secure channel.
+ - The plaintext of DK will never be exposed outside of the enclaves.
+
+---
+
+## Build Instructions
+For more details please refer to [build-instructions](https://github.com/intel/ehsm/blob/bad3ba0525fa1c03f1cab55c48288df26df877af/docs/build-instructions.md).
```

Comment:
you can move the build-instructions.md to CCzoo as well

Reply:
As we aligned this morning, ehsm is still in an active development stage. Build and deployment instructions are frequently changed. We preferred to moving to cczoo once they are stable.

Reply:
Ok, got it, Thanks